kong/clustering/version_negotiation/init.lua

```diff
@@ -0,0 +1,355 @@
+
+
+local _M = {}
+
+local function validate_connection_type()
+  if ngx.req.get_method() ~= "POST" then
+    return nil, "INVALID METHOD"
+  end
+
+  if ngx.var.http_content_type ~= "application/json" then
+    return nil, "Invalid Content-Type"
+  end
+
+  return true
+end
+
+local function get_body()
+  ngx.req.read_body()
+  local body = ngx.req.get_body_data()
+  if body then
+    return body
+  end
+
+  local fname = ngx.req.get_body_file()
+  if fname then
+    return pl_file.read(fname)
+  end
+
+  return ""
+end
+
+local function response(status, body)
+  ngx.status = status
+
+  if type(body) == "table" then
+    ngx.header["Content-Type"] = "application/json"
+    body = cjson.encode(body)
+  end
+
+  ngx.say(body)
+  return ngx.exit(status)
+end
+
+local function response_err(msg)
+  return response(400, { message = msg })
+end
+
+local function verify_request(body)
+  if type(body.node) ~= "table" then
+    return false, "field \"node\" must be an object."
+  end
+
+  if type(body.node.id) ~= "string" then
+    return false, "field \"node.id\" must be a string."
+  end
+
+  if type(body.node.type) ~= "string" then
+    return false, "field \"node.type\" must be a string."
+  end
+
+  if type(body.node.version) ~= "string" then
+    return false, "field \"node.version\" must be a string."
+  end
+
+  if type(body.services_requested) ~= "table" then
+    return false, "field \"services_requested\" must be an array."
+  end
+
+  return true
+end
+
+
+local function node_info()
+  return {
+    id = kong.node.get_id()
+  }
+end
+
+local function cp_priority(name, req_versions, known_versions)
+  local versions_set = {}
+  for _, version in ipairs(req_versions) do
+    versions_set[version] = true
+  end
+
+  for _, v in ipairs(known_versions) do
+    if versions_set[v.version] then
+      return true, {
+        name = name,
+        version = v.version,
+        message = v.message,
+      }
+    end
+  end
+
+  return false, { name = name, message = "No valid version" }
+end
+
+local all_known_services = require "kong.clustering.version_negotiation.services_known"
+
+local function check_node_compatibility(client_node)
+  if client_node.type ~= "KONG" then
+    return nil, ("unknown node type %q"):format(client_node.type), CLUSTERING_SYNC_STATUS.UNKNOWN
+  end
+
+  return clustering_utils.check_kong_version_compatibility(KONG_VERSION, client_node.version)
+end
+
+local function do_negotiation(req_body)
+  local services_accepted = {}
+  local services_rejected = {}
+
+  for i, req_service in ipairs(req_body.services_requested) do
+    if type(req_service) ~= "table" or type(req_service.name) ~= "string" then
+      return nil, "Malformed service requested item #" .. tostring(i)
+    end
+
+    local name = req_service.name
+
+    if type(req_service.versions) ~= "table" then
+      table.insert(services_rejected, {
+        name = name,
+        message = "invalid \"versions\" table.",
+      })
+      goto continue
+    end
```

Comment:
I'd be inclined to send back a 400 (with no negotiation whatsoever) if the version is not an array.

Reply:
done